### ISSUE TITLE
Babies often trigger false positives

### ISSUE THREAD

**Issue by wingman-jr-addon** [wingman-jr-addon/wingman_jr#22]
Need to add more babies into the training data!

**Comment by TechnikEmpire** [comment]
That won't work. You need to train an object detector rather than full-context binary classification.

**Comment by wingman-jr-addon** [comment]
I've been sticking with the full-context approach because an object detector would be unlikely to catch some of the types of NSFW things I'm looking for that are almost entirely relative location-based - this is similar to the types of issues around random crop. For example, a photo of a volleyball player being SFW or NSFW may be primarily in the framing of the photo. I think what could maybe work someday would be to do an object detector with an additional layer looking at relative contexts of the boxes? But for now I'm running MobileNetV2-based network.

Babies do have certain interesting features that are rather unlike adults or even children. Their facial structures are different, their skin is  more rounded, and their typical photo contexts often resemble unsafe situations if you squint hard enough. I'll see how far I can get with the full-context approach.

**Comment by TechnikEmpire** [comment]
I'll let you know what kind of precision I get when I can.

**Comment by wingman-jr-addon** [comment]
Ok, with the recent update to model SQRX62 I'm planning to close this. To resolve this, I added several thousand baby, pregnancy, and family themed photos for hard negatives. Additionally, I added a significant number of pregnancy themed NSFW images into the mix. Pregnant bellies and breastfeeding photos are treated on more of a case-by-case ranging from fine to NSFW. Overall results have a significant reduction in baby photos and family photos triggering false positives with more mixed results on pregnant belly photos.

**Comment by TechnikEmpire** [comment]
Are you doing transfer learning? I'm trying to train a binary classifier from scratch (mobilenet v2) and despite having millions of images that are all correctly labeled, I can't get the model to converge. In fact after several hours of training the loss skyrockets. Thanks for any input you have. I am waiting on fine tuning right now to see if it works but I thought I'd ask.

**Comment by wingman-jr-addon** [comment]
Yes, my released model is transfer learning based on 224x224 MobileNet V2 from TF 2.0's ImageNet tranined one. However, I have also been working on training from scratch as my dataset has grown and it's getting closer to accuracy parity. (I'd like to get to a point of training from scratch so that I can use custom sizes e.g. 300x300.) Training from scratch has generally converged, but it did suffer from more problems when my dataset was smaller and in particular when my LR was too low. Training from scratch seems to work better with e.g. Adam = 0.005 instead of something in the ballpark of Adam = 0.00003 for finetuning. I still haven't done too much training from scratch yet; I only have a TX1 for training so my experimentation is limited.

In both methods, the absolute accuracy is still poorer than I would like, but  the generalization is better than accuracy alone would indicate because I use 4 classes instead of a binary classifier (safe, questionable, racy, explicit) along with a custom loss function that penalizes in an asymmetric, gradated way. That may be helping to train to convergence?

**Comment by TechnikEmpire** [comment]
Thanks for sharing. Yeah I was going to use 6 classes, one of which is plain images or "absolute negative". I suppose to don't need that negative class.I've managed to transfer learn since my last comment and accuracy seems decent. Validation set reports ~97% but I don't trust that. I'm going to test on brand new material tonight and compare to my squeezenet model.I had a hell of a time getting the model into opencv dnn so I'm just getting to manual validation now.I wonder if we can collaborate. What I have is 3D scans of people (that I cannot release obviously) but the benefit is that, with a blender plugin I've written, I can generate essentially limitless, high quality data. The plugin generates cameras from your current POV that spiral around your target and then renders all cameras to file along with the bounding box of the original camera's target.Then you apply something like imgaug (on github) and you can easily wind up with millions of images per class.I had planned to train a SSD but I need model execution to do real time cinema on average PCs so I've shelved that training set.I just need to figure out how to get TF to train my multi-class set without it flying off the handle and reporting 80,000+ loss values after an hour lol.On Nov 24, 2019 10:22 PM, wingman-jr-addon <notifications@github.com> wrote:Yes, my released model is transfer learning based on 224x224 MobileNet V2 from TF 2.0's ImageNet tranined one. However, I have also been working on training from scratch as my dataset has grown and it's getting closer to accuracy parity. (I'd like to get to a point of training from scratch so that I can use custom sizes e.g. 300x300.) Training from scratch has generally converged, but it did suffer from more problems when my dataset was smaller and in particular when my LR was too low. Training from scratch seems to work better with e.g. Adam = 0.005 instead of something in the ballpark of Adam = 0.00003 for finetuning. I still haven't done too much training from scratch yet; I only have a TX1 for training so my experimentation is limited.
In both methods, the absolute accuracy is still poorer than I would like, but  the generalization is better than accuracy alone would indicate because I use 4 classes instead of a binary classifier (safe, questionable, racy, explicit) along with a custom loss function that penalizes in an asymmetric, gradated way. That may be helping to train to convergence?
